### ISSUE TITLE
File descriptor leak checking

### ISSUE THREAD

**Issue by daurnimator** [ziglang/zig#6152]
I've found some resource leaks in code due to missing `file.close()` calls.
As a debug feature, we should be able to detect this and at least throw an error in tests.

Implementation ideas:
  - (debug mode only) global linked list of file descriptors: objects like `std.fs.File` could add themselves to the list on open and remove on `.close()`
  - (release-safe mode and debug mode): global atomic counter incremented on `open` and decremented on `close`. Check for equality at start/end of tests.

**Comment by kubkon** [comment]
If we could pull this off, that would be insanely useful IMHO. :+1:

**Comment by zigazeljko** [comment]
Would this only apply to `std.fs.File` or also to `std.os`?

**Comment by ArthurBrussee** [comment]
I agree this would be very useful! The demo of an allocator detecting a leak was amazing.

I wonder if there could be a generic 'sentinel' mechanism that could be added to anything that is meant to be freed/released/... as I'd like to use this on user types as well.

**Comment by mrakh** [comment]
FYI, you can already do this with `valgrind` using the `--track-fds=yes` option.

**Comment by arqv** [comment]
I think the purpose of doing this inside the standard library is the same that motivated the creation of the GPA having a leak detector built-in, so it would be nice to have it done without the need of any external tools.

**Comment by matu3ba** [comment]
Subprocess creation should also not leak file descriptors, as they can pose a security thread ie in chrooted environments.

[This guide](https://docs.fedoraproject.org/en-US/Fedora_Security_Team/1/html/Defensive_Coding/sect-Defensive_Coding-Tasks-Descriptors-Child_Processes.html)
explains 2 possible solutions:
1. treat every file descriptor special with `*CLOEXEC` during opening to let the Kernel know that it should not clone the file descriptor on subprocess creation
2. all file descriptors, which the child process will not need, are closed

Option 1: I dont know, if option 1 has a significant performance impact upon `clone` as the Kernel may need to traverse some objects. It could be more tedious to make this a default, but on the other hand prevents any possible race conditions.
Option 2: "Traditionally, this was implemented as a loop over file descriptors ranging from 3 to 255 and later 1023. But this is only an approximatio because it is possible to create file descriptors outside this range easily". The potential problem with this approach is explained [here](https://docs.fedoraproject.org/en-US/Fedora_Security_Team/1/html/Defensive_Coding/sect-Defensive_Coding-Tasks-Descriptors-Limit.html): "Even though this algorithm is racy in the sense that the FD_SETSIZE first descriptors could fill up, a very high degree of physical parallelism is required before this becomes a problem."

**Comment by matu3ba** [comment]
For my comment I am assuming that we want to leak-check arbitrary handles / file descriptors.

The main problem of leak detection in general is that observing via memory structure synchronization and IPC, which can alter behavior (unless traced by Kernel).

1. Observing file descriptor leakage in Valgrind prevented me to get msync synchronization problems on Linux here https://github.com/ziglang/zig/pull/13639#issuecomment-1325637164. However, doing something similar sounds like the most portable option.
2. Observing the process state via debugger api requires more expensive Kernel calls, but has less effect on observed process (if we do the check in the parent process after spawn is completed and the child running).
3. Using tracing of programs via Kernel sounds to me like the best option, although memory costs can be high.
4. Only the existence of leaks (without tracing its origin) can be also done in the exit code via the process inspecting its own in-Kernel structures (`/proc/`). However, I'm not sure if all Kernels support that.

I would recommend looking into this again once Zig has incremental compilation, since that requires several debug bindings and should provide a better understanding of the necessary complexity for 2.
So far I dont see the significant advantage on recreating valgrind, if also there exist the less intrusive methods 2+3.

> (debug mode only): global var containing array of known file descriptors -> stack frames. On exit iterate the array and show tracebacks for which fds were leaked
> (release-safe mode and debug mode): global atomic counter incremented on open and decremented on close. Check for equality at start/end of tests.

The main problem with these approaches is that some leaks are racy, since for example in Linux after fork CLOEXEC can only be applied in user-space in a racy manner, because the setup code in user-space must prepare the child process after `clone()` before doing the `execve`.
If we do full-stop locks to synchronizes or we write to a pipe etc, we will alter the behavior. This could be fine, if we control the process scheduler, but without it can be painful to reproduce behavior.
